internal/rust/semantic/src/types.rs

```diff
@@ -0,0 +1,705 @@
+use std::{cmp, collections::HashMap, fmt};
+
+// PolyType represents a generic parametrized type.
+//
+// TODO:
+//     Do not derive PartialEq implementation.
+//     Instead provide a custom implementation
+//     that instantiates both polytypes with the
+//     same type variables.
+//
+//     Note this requires a substitution, so remove
+//     this derivation once substitutions are defined.
+//
+#[derive(Debug, Clone, PartialEq)]
+pub struct PolyType {
+    pub free: Vec<Tvar>,
+    pub bnds: Option<HashMap<Tvar, Kind>>,
+    pub expr: MonoType,
+}
+
+impl fmt::Display for PolyType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match &self.bnds {
+            Some(bnds) => {
+                let mut bounds = Vec::new();
+                for tv in &self.free {
+                    if let Some(kind) = bnds.get(tv) {
+                        bounds.push(BoundTvar {
```

Comment:
I guess you could avoid building up a new vector to hold those variables and directly push the result of formatting in the final string, that should use less memory and blablabla.

But I guess this is only for debugging and does not need to be very efficient